timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/redis/RedisService.java

```diff
@@ -0,0 +1,305 @@
+                    runWithRetries(() -> pipelined.setex(id.getBytes(), redisTtlInSeconds, taskByteArr), "SETEX");
+                } catch (Exception e) {
+                    allPushed = false;
+                    LOG.error("Error pushing id " + id + " to Redis.", e);
+                }
+            }
+        }
+        return allPushed;
+    }
+
+    public <T> boolean pushToRedisHash(Map<String, T> idsToValuesMap) {
+        return pushToRedisHash(idsToValuesMap, redisTtlInSeconds);
+    }
+
+    public <T> boolean pushToRedisList(String listName, T value) {
+        boolean allPushed = true;
+        try (Jedis jedis = jedisPool.getResource()) {
+            try {
+                byte[] valueByteArr = getBytes(value);
+                runWithRetries(() -> jedis.rpush(listName.getBytes(), valueByteArr), "RPUSH");
+            } catch (Exception e) {
+                allPushed = false;
+                LOG.error("Error pushing item to Redis " + listName + " list", e);
+            }
+        }
+        return allPushed;
+    }
+
+    public long getListLength(String listName) {
+        try (Jedis jedis = jedisPool.getResource()) {
+            try {
+                return runWithRetries(() -> jedis.llen(listName.getBytes()), "LLEN").longValue();
+            } catch (Exception e) {
+                LOG.error("Error returning Redis " + listName + " list length", e);
+                return -1;
+            }
+        }
+    }
+
+    public  <T> List<T> getRangeFromRedisList(String listName, int start, int end) {
+        List<T> ids = new ArrayList<>();
+        try (Jedis jedis = jedisPool.getResource()) {
+            List<byte[]> serializedObjects = runWithRetries(() -> jedis.lrange(listName.getBytes(), start, end), "LRANGE");
+            for (int i = 0; i < serializedObjects.size(); i++) {
+                byte[] serializedObject = serializedObjects.get(i);
+
+                if (serializedObject == null || serializedObject.length == 0) {
+                    continue;
+                }
+                Kryo kryo = kryoPool.obtain();
+                try {
+                    T object = (T) kryo.readClassAndObject(new Input(serializedObject));
+                    ids.add(object);
+                } catch (Exception e) {
+                    LOG.error("Error getting elements from Redis " + listName + " list", e);
+                } finally {
+                    kryoPool.free(kryo);
+                }
+
+            }
+        } catch (Exception e) {
+            LOG.error("Error getting elements from Redis " + listName + " list", e);
+            ids = new ArrayList<>();
+        }
+        return ids;
+    }
+
+    public List<String> popElementsFromRedisList(String listName, int amount) {
+        lock();
+        List<String> elements = getRangeFromRedisList(listName, 0, amount - 1);
+        trimRedisList(listName, elements.size(), -1);
+        release();
+        return elements;
+    }
```

Comment:
there is no built-in multiple POP for Redis list and issuing multiple POP requests to the Redis server seems to me too expensive,  soI implemented pop by get & cut (saw recommendations in StackOverflow)

Reply:
btw
I see our default limit for the number of fetched bulks in one call to fetchAndDelete is 10
so it's not too much.. but we can (or should?) possibly increase it

Reply:
@edengal what about a sorted set? it has SPOP too

Reply:
Sure, use ZADD with a score representing the time and ZPOPMIN to get the earliest, it a bit slower but still pretty fast.
@oz-zafar